FIVETRAN_CODING_PRINCIPLES.md

```diff
@@ -0,0 +1,192 @@
+- Feel free to use GitHub Copilot as a reviewer to catch basic issues early.
+
+4. Add comments in the diff to highlight important logic or non-obvious changes.
+
+5. Respond professionally to reviewer comments. Remember we are here to help and will strive to ask for only necessary changes to future proof your contribution and make it understandable to others.
+
+---
+
+## 3. Preparing code for PR submission
+
+To adhere to Fivetran engineering standards, follow this process:
+
+1. Write your `connector.py` solution (start from system-generated baselines if applicable).
+2. Feed it to the cursor agent `connector_sdk_template` to generate `connector_test.py` following best practices.
+3. Review generated revisions (e.g., ensure `fivetran debug` works).
+4. Feed the project to cursor agent `connector_sdk_readme` to generate a consistent `README.md`.
+5. Submit the PR.
+
+---
+
+## 4. Example connector guidelines
+
+### 4.1 Template connector
+
+A reference implementation is available here:  
+🔗 [Fivetran Connector SDK Template Example](https://github.com/fivetran/fivetran_connector_sdk/tree/main/template_example_connector)
+
+This includes:
+- A `connector.py` template.
+- A `README.md` template illustrating recommended structure and conventions.
+
+This is a great place to start to learn our coding practices through templated examples.
+
+
+### 4.2 General best practices
+
+- Import only necessary modules and libraries.
+- Use clear, consistent, descriptive names for functions and variables.
+- Use UPPERCASE_WITH_UNDERSCORES for constants (e.g., `CHECKPOINT_INTERVAL`).
+- Keep constants private by default (e.g., `__CHECKPOINT_INTERVAL`).
+- Add docstrings explaining each function’s purpose.
+- Add comments for complex logic and user customization points.
+- Split large functions into smaller helper functions for readability.
+- Use logging judiciously — avoid excessive or redundant logs.
+- Implement error handling that targets specific exceptions.
+- Add retry logic (with exponential backoff) for API requests.
+- Define a complete data model in the `schema()` function, with primary keys and data types.
+- Avoid loading all data into memory — use pagination or streaming.
+- Document pagination, upsert/update/delete, and checkpointing with clear comments.
+- Checkpoint state regularly to resume from the last successful sync and cleanly send data to your warehouse.
+- Refer to the [Connector SDK Best Practices](https://fivetran.com/docs/connectors/connector-sdk/best-practices).
+
+---
+
+## 5. Example development guidelines
+
+### 5.1 Example principles
+
+- Self-explanatory and well-commented code.
+- Must produce working code (can require the SDK Playground).
+- No identifiable customer information should be included.
+
+
+
+### 5.2 Demonstrate one core concept
+
+- Each example should focus on one SDK feature or pattern (e.g., incremental sync, schema discovery, error handling).
+- Split complex connectors into multiple examples.
```

Comment:
@mesmith027 I am not sure I am following what it means. A complex connector is one that has many features. For source connectors, it's unfeasible to "split", a connector works as a whole and single entity. For pattern/feature implementation examples, it's an inherent and obvious core principle to implement that very feature and nothing else.